ldap/servers/slapd/pagedresults.c

```diff
@@ -1018,10 +1018,10 @@ pagedresults_lock(Connection *conn, int index)
     }
     pthread_mutex_lock(pageresult_lock_get_addr(conn));
     prp = conn->c_pagedresults.prl_list + index;
-    pthread_mutex_unlock(pageresult_lock_get_addr(conn));
     if (prp->pr_mutex) {
         PR_Lock(prp->pr_mutex);
     }
+    pthread_mutex_unlock(pageresult_lock_get_addr(conn));
```

Comment:
A closer look confirms my unease about that one:
There is a deadlock risk:
  The page search per connection mutex is now locked then the prp mutex
  but op_shared_search locks them in the opposite order:
  	    pagedresults_lock(pb_conn, pr_idx);                   ===> The prp mutex 
            pthread_mutex_lock(pagedresults_mutex);        ===> The per connection mutex

Reply:
I'm not seeing this.  pagedresults_lock() releases the per_conn lock before it returns.  So you have to call pagedresults_lock() before taking the per_conn mutex.  Everywhere I checked this is the order that is used:

pagedresults_lock(conn, idx)
  --> pthread_mutex_lock(pagedresults_mutex);
  <-- pthread_mutex_unlock(pagedresults_mutex);
pagedresults_unlock(conn, idx)

Yes pagedresults_lock/pagedresults_unlock takes the per_conn mutex while it locks/unlock the prp_mutex, but it shouldn't deadlock.  I'm probably missing the scenario you are worried about, can you point out where this is happening?

Reply:
Ok I see how it can happen.  Thread 1 has prp lock, but doesn't have the per_conn lock yet, thread 2 takes per_conn lock (via pagedresults_lock()) as it tries to get the prp lock.  And deadlock.  Ugh, too bad :-(